README.md
Analog Devices AD56X4 12/14/16-Bit Quad Channel DAC Arduino Library
===================================================================

An [Arduino](http://arduino.cc) library to control an Analog Devices [AD56X4](http://www.analog.com/en/digital-to-analog-converters/da-converters/ad5624r/products/product.html?doc=AD5624R_5644R_5664R.PDF) series 12/14/16-bit Quad channel DAC (Digital to Analog Converter) by SPI. This library interfaces the following chips

* AD5624  : 12-bit
* AD5664  : 16-bit
* AD5624R : 12-bit with an internal voltage reference that can be used
* AD5644R : 14-bit with an internal voltage reference that can be used
* AD5664R : 16-bit with an internal voltage reference that can be used


Copyright (c) 2013, Freja Nordsiek

This software is distributed under a [2-Clause BSD License](./COPYING.txt).



Making The Library Package
--------------------------

On a system with `make`, the provided [Makefile](./Makefile) can be used to build a `.zip` file containing the whole library by simply running `make` or `make package`. The package will be named `AD65X4_VERSION.zip` where `VERSION` is the current package [version](./VERSION.txt).



Hardware Information
--------------------

All the AD56X4 series are connected in a similar way in order for the Arduino to communicate with it by SPI (Serial Peripheral Interface). The DIN pin on the chip (pin 8 on the MSOP and LFCSP packages and pin C1 on the WLCSP package) corresponds to the Arduino's MOSI pin (pin 11 on the Uno, pin 51 on the Mega, and ICSP-4 on all that have the ICSP header). The SCLK pin on the chip (pin 7 on the MSOP and LFCSP packages and pin D1 on the WLCSP package) corresponds to the Arduino's SCK pin (pin 13 on the Uno, pin 52 on the Mega, and ICSP-3 on all that have the ICSP header). The SYNC pin on the chip (pin 6 on the MSOP and LFCSP packages and pin D2 on the WLCSP package) is the Slave Select (SS) pin, which might as well be the Arduino's SS pin (pin 10 on the Uno and pin 53 on the Mega). The SS pin to use has to be given as the first argument of every public function in this library (makes it easier to have multiple SPI devices with different SS pins).

The chip can receive data by SPI at speeds up to 50 MHz, which is significantly greater than the Uno and Mega can do at the present time (8 MHz as of 2013-08-20). So, the SPI clock divider setting doesn't really matter except for how long it will take for commands to complete.

The chips all use a voltage reference that is the upper rail for the analog outputs and determines the resolution of the output voltage. All the chips can take an external voltage reference (pin 16 on the MSOP and LFCSP and pin A1 on the WFCSP). The AD56X4R series have an internal voltage reference that can optionally be used instead (off by default). It is 1.25 V for the AD56X4R-3 series and 2.5 V for the AD56X4R-5 series.



Using The Library
-----------------

An [Arduino](http://arduino.cc) program using this library must first include the library header in addition the the SPI library header.

```Arduino
#include <SPI.h>
#include <AD56X4.h>
```

All library functions require that the Arduino pin that is the Slave Select for the AD654X (SYNC pin on the device) be given as the first argument. Hence, the calling format looks like

```Arduino
AD56X4.function(AD56X4_SS_pin, ...);
```

When initializing, the Arduino's Slave Select pin (10) must be set to OUTPUT in addition to whatever pin is used for the Slave Select of the AD56X4 device (pin 10 is a good candidate to avoid wasting a pin). As the AD56X4 device can work with up to 50 MHz SPI messages while the Arduino's maximum speed at the time of this writing (2013-08-20) is 8 MHz, the particular clock divider setting doesn't really matter, though setting it as fast as possible to `SPI_CLOCK_DIV2` will reduce the time it takes to give the AD56X4 commands. Unless it is desirable to keep the previous state of the AD65X4 on Arduino power up or reset, one should do a full reset of the device with `AD56X4.reset(AD56X4_SS_pin, true)`. An initialization sequence would look like

```Arduino
pinMode(10,OUTPUT);
pinMode(AD56X4_SS_pin,OUTPUT);
SPI.setClockDivider(SPI_CLOCK_DIV2);
SPI.begin();

AD56X4.reset(AD56X4_SS_pin,true);
```

Note, when any library function is called, the [SPI Bit Order](http://arduino.cc/en/Reference/SPISetBitOrder) is set to `MSBFIRST` and the [SPI Data Mode](http://arduino.cc/en/Reference/SPISetDataMode) is set to `SPI_MODE1`. These are not changed back at the end of the function call, so they will need to set before using the SPI bus with another device.

The AD56X4 series DACs all have the value they are outputting and a buffer holding the next value to output. These are referred to as the DAC and input registers respectively. Setting the input registers (function `AD56X4.setChannel`) does not change the analog voltages on the channels unless the AD56X4 is told to update the output (DAC register) at the same time (optional set modes `AD56X4_SETMODE_INPUT_DAC` or `AD56X4_SETMODE_INPUT_DAC_ALL`) or that is the default setting for the channel (function `AD56X4.setInputMode`). The function `AD56X4.updateChannel` is used to update the output (DAC register) to the current value of the input register.



Example Code
------------

An example program, [Four_Sine_Waves](./examples/Four_Sine_Waves/Four_Sine_Waves.ino), is provided that writes 10 Hz sine waves to each channel that are out of phase with each other in 90 degree increments.



Library Functions
-----------------

*   ```Arduino
    void AD56X4.setChannel(int SS_pin, byte setMode, byte channel, word value)
    void AD56X4.setChannel(int SS_pin, byte setMode, word values[])
    void AD56X4.setChannel(int SS_pin, byte setMode, word value_D, word value_C, word value_B, word value_A)
    ```
    
    Sets the value/s of the specified channel/s on the chip (Slave Select pin `SS_pin`). Values (last argument/s) have to be `word` types (unsigned 16-bit integers, which are `unsigned int` as well). For 12 and 14-bit chips, the last 4 and 2 bits respectively are ignored. `setMode` specifies how the registers are set when the channels/s are set, and the valid values are listed below
    
    *   `AD56X4_SETMODE_INPUT`          Set only the input register (doesn't change voltage output).
    *   `AD56X4_SETMODE_INPUT_DAC`      Set both the input and DAC registers (voltage output is updated).
    *   `AD56X4_SETMODE_INPUT_DAC_ALL`  Set the input register for that channel and cause all input registers to then be copied into their respective DAC registers (output updated on all channels).
    
    For the first overload (calling method), only one channel (or all channels) are set to `value` (or all channels set to it). The valid values of `channel` are
    
    *   `AD56X4_CHANNEL_A`    Channel A only.
    *   `AD56X4_CHANNEL_B`    Channel B only.
    *   `AD56X4_CHANNEL_C`    Channel C only.
    *   `AD56X4_CHANNEL_D`    Channel D only.
    *   `AD56X4_CHANNEL_ALL`  All channels.
    
    They CANNOT be bitwise OR'ed together. In the second and third calling overloads, each channel is set to the given values, which can either be a 4-element array (channel D to A order) or four separate arguments.

*   ```Arduino
    void AD56X4.updateChannel(int SS_pin, byte channel)
    ````
    
    For the chip with the Slave Select pin `SS_pin`, the DAC register (and therefore voltage output) of the specified channel (see `setChannel` above for how channels are specified) is set to the value of its input register. If `setChannel` was called with `setMode = AD56X4_SETMODE_INPUT`, then this function will have to be called on the channel in order for the output voltage to be updated.

*   ```Arduino
    void AD56X4.powerUpDown(int SS_pin, byte powerMode, boolean channels[])
    void AD56X4.powerUpDown(int SS_pin, byte powerMode, boolean channel_D, boolean channel_C, boolean channel_B, boolean channel_A)
    void AD56X4.powerUpDown(int SS_pin, byte powerModes[])
    ```
    
    For the chip with the Slave Select pin `SS_pin`, the power state/mode of the channels are set. The valid power mode/s are
    
    *   `AD56X4_POWERMODE_NORMAL`          On.
    *   `AD56X4_POWERMODE_POWERDOWN_1K`    Off and connected by a 1k resistor to ground.
    *   `AD56X4_POWERMODE_POWERDOWN_100K`  Off and connected by a 100k resistor to ground.
    *   `AD56X4_POWERMODE_TRISTATE`        Off in a tri-state operation.
    
    For the first and second overloads (function calling methods), one power mode `powerMode` is applied to the specified channels. The channels are specified either as a 4-element `boolean` array (channel D to A order) or as four `boolean` arguments. `true` means set to the given power mode and `false` means not (keep current power mode). The third overload sets the power mode of each channel to the power modes given in the 4-element array `powerModes[]` (channel D to A order).

*   ```Arduino    
    void AD56X4.reset(int SS_pin, boolean fullReset)
    ```
    
    Resets the AD56X4 chip with Slave Select pin `SS_pin`. The DAC and input registers are all reset to 0, and thus the analog voltages are set to zero. If `fullReset` is `true`, then a full reset is done which resets the channel powerMode to on (`AD56X4_POWERMODE_NORMAL`), the external voltage reference is used (internal reference, if present, turned off), and the input mode (LDAC register in the chip data sheet) is reset (updating the input registers does not cause automatic updates of the DAC registers from their values). 

*   ```Arduino
    void AD56X4.setInputMode(int SS_pin, boolean channels[])
    void AD56X4.setInputMode(int SS_pin, boolean channel_D, boolean channel_C, boolean channel_B, boolean channel_A)
    ```
    
    Sets the input modes of each channel on the chip (Slave Select pin `SS_pi`). The modes are `boolean` with `true` meaning that setting the input register automatically sets the DAC register (and therefore voltage output) to the same value and `false` meaning no auto update of the DAC register. The modes for each channel can either be given by a 4-element `boolean` array (channel D to A order) or as four separate input arguments.

*   ```Arduino
    void AD56X4.useInternalReference (int SS_pin, boolean yesno)
    ```
    
    Choose whether the chip (Slave Select pin `SS_pin`) uses the internal voltage reference (`yesno = true`) or the external reference pin (`yesno = false`). Only applicable for the chips that have an internal reference (AD56XR).


ChangeLog.txt
2013-12-20 Freja Nordsiek
	* Changed license from 3-clause BSD to 2-clause BSD.
	* Renamed LICENSE.txt to COPYING.txt
	* Incremented version to 0.1.1 due to license change.

2013-08-28 Freja Nordsiek
	* Added "AD56X4." before function declarations in README.md.
	* Incremented version to 0.1 since the library appears to work.

2013-08-26 Freja Nordsiek
	* Fixed lack of time evolution bug in Four_Sine_Waves example.
	* Changed sine wave frequency to 10 Hz in Four_Sine_Waves
	  example.

2013-08-20 Freja Nordsiek
	* Added hardware documentation to AD56X4.* and README.md and
	  library documentation to README.md.

2013-08-19 Freja Nordsiek
	* Changed Four_Sine_Waves.ino to use builtin value of pi.
	* Made another powerUpDown overload call the private one.

2013-08-18 Freja Nordsiek
	* Hard wrapped function declaration lines that were too long.
	* Changed the private overloads of powerUpDown and setInputMode to
	  inline and made the public overloads call them. 

2013-08-17 Freja Nordsiek
	* Initial version.
	* Added example: examples/Four_Sine_Waves/Four_Sine_Waves.ino
	* Slightly modified license to original 3-clause BSD.
	* Added README.md


AD56X4.cpp
/* Copyright (c) 2013, Freja Nordsiek
   All rights reserved.
   
   Redistribution and use in source and binary forms, with or
   without modification, are permitted provided that the
   following conditions are met:
   
   1. Redistributions of source code must retain the above
   copyright notice, this list of conditions and the following disclaimer.
   
   2. Redistributions in binary form must reproduce the above
   copyright notice, this list of conditions and the following
   disclaimer in the documentation and/or other materials
   provided with the distribution.
   
   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
   NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
   LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
   HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
   CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
   OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
   SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

/*
   AD56X4.cpp: Library for controlling the Analog Devices AD56X4
                 Quad DAC family:
                   AD5624   12-bit
                   AD5664   16-bit
                   AD5624R  12-bit with internal reference
                   AD5644R  14-bit with internal reference
                   AD5664R  16-bit with internal reference
             Data sheet can be found at http://www.analog.com/en/digital-to-analog-converters/da-converters/ad5624r/products/product.html?doc=AD5624R_5644R_5664R.PDF
   
   Author:   Freja Nordsiek
   Notes:
   History:  * 2013-08-15 Created.
*/

/* http://www.analog.com/en/digital-to-analog-converters/da-converters/ad5624r/products/product.html?doc=AD5624R_5644R_5664R.PDF

   The Analog Devices AD56X4 are Quad channel 1X-bit Digital to
   Analog Converter (DAC) controlled by SPI. It can use an external
   reference or its own internal reference if present (an R is
   appended to the chip name) and has various shutdown modes for
   each channel. Each channel has separate DAC (output) and input
   (buffer) registers and commands exist to set just the input
   register, assign the input to the DAC, or set both.
   
   The chip is an SPI slave that expects SPI_MODE1 for its polarity
   and phase (read on falling edge and clock low while idle) that
   can operate at up to 50 MHz, which is far above what the arduino
   can do (8 MHz for the Uno). The DIN, SCLK, and SYNC pins in the
   chip documentation correspond to the Arduino MOSI, SCK, and SS
   pins (doesn't have to use the SS for the arduino, but it is a
   a reasonable default pin). Messages must be sent MSB first and
   are 24 bits. The first two bits are junk and are ignored. The
   next three bits are the command bits, which tell the DAC what
   operation to do. The next three bits are the address bits, which
   tells what channel (or all) should be set. The last two whole
   bytes are a word specifying the value to set the appropriate
   register on the DAC channel if setting a channel, or other needed
   information if doing another sort of command. For the 14-bit and
   12-bit chips, the last 2 and 4 bits respectively are ignored if
   an input or DAC register is being set. The command bits and address
   bits are
   
   Command bits
   | C2 | C1 | C0 | Command
   ------------------------------------------------------------
   | 0  | 0  | 0  | Write to input register N
   | 0  | 0  | 1  | Update DAC register N
   | 0  | 1  | 0  | Write to input register N and update on all
   | 0  | 1  | 1  | Write to and update channel N
   | 1  | 0  | 0  | Power up-down
   | 1  | 0  | 1  | Reset
   | 1  | 1  | 0  | Set LDAC register
   | 1  | 1  | 1  | Internal reference on/off (if present)
   ------------------------------------------------------------
   
   Address bits
   | A2 | A1 | A0 | Channel/s
   --------------------------
   | 0  | 0  | 0  | A
   | 0  | 0  | 1  | B
   | 0  | 1  | 0  | C
   | 0  | 1  | 1  | D
   | 1  | 1  | 1  | All
   --------------------------
   
   For the power up-down command, the power mode specified with
   data bits DB5 and DB4 are applied to the channels specified by
   ones in the bit mask made by DB3 to DB0 (in channel D to A
   order). The power modes are
   
   Power up-down mode
   | DB5 | DB4 | mode
   ------------------------------
   |  0  |  0  | normal operation
   |  0  |  1  | 1k to ground
   |  1  |  0  | 100k to ground
   |  1  |  1  | tri-state
   ------------------------------
   
   For the reset command, DB0 controls exactly what is reset.
   Regardless of its value, the input and DAC registers are all
   reset to zero. If DB0 is set to one; then the LDAC register
   is reset to zero, all channels powered up to normal mode, and
   the internal reference turned off (means external reference is
   used).
   
   For the set LDAC register command, whether to just set the DAC
   register to the input register every time it is changed or not,
   is set for each channel with DB3 to DB0 corresponding to channel
   D through A. A zero means that setting the input register does
   not automatically set the DAC register (value isn't written out
   as output yet) and a one means that the DAC register (and hence
   the output) is set the the input register the moment the input
   register is changed.
   
   For the internal reference on/off command, setting DB0 to zero
   turns it off and setting it to one turns it on. When off, the
   external reference is used as the voltage reference. This
   command only works in the R versions that have an internal
   reference.
   
*/

#include "Arduino.h"
#include <SPI.h>
#include <AD56X4.h>

AD56X4Class AD56X4;

/* Commands the AD564X DAC whose Slave Select pin is SS_pin to
   set the values of the specified channel/s. The values are word
   with the 12/14/16-bit values the channels should be set at (last
   4 and 2 bits are ignored for 12-bit and 14-bit DACs). Either
   one channel can be set (or all channels set to the same value)
   with a channel byte that must be one of
   
   AD56X4_CHANNEL_A
   AD56X4_CHANNEL_B
   AD56X4_CHANNEL_C
   AD56X4_CHANNEL_D
   AD56X4_CHANNEL_ALL
   
   or separate values (as an array in D to A order or as four
   separate arguments) be set to each one. There are three ways
   they can be set, or set modes. They are
   
   AD56X4_SETMODE_INPUT          Set the channel/s's input register.
   AD56X4_SETMODE_INPUT_DAC      Set both the input and DAC
                                   registers for the channel/s.
   AD56X4_SETMODE_INPUT_DAC_ALL  Set channel/s's input register and
                                   then update all DAC registers
                                   from the input registers.
*/
void AD56X4Class::setChannel (int SS_pin, byte setMode, byte channel,
                              word value)
{
  // Don't do anything if we weren't given a valid setMode.
  if (setMode == AD56X4_SETMODE_INPUT 
      || setMode == AD56X4_SETMODE_INPUT_DAC 
      || setMode == AD56X4_SETMODE_INPUT_DAC_ALL)
    AD56X4.writeMessage(SS_pin,setMode,channel,value);
}
void AD56X4Class::setChannel (int SS_pin, byte setMode, word values[])
{
  // Don't do anything if we weren't given a valid setMode.
  if (setMode == AD56X4_SETMODE_INPUT 
      || setMode == AD56X4_SETMODE_INPUT_DAC 
      || setMode == AD56X4_SETMODE_INPUT_DAC_ALL)
    {
      // It luckily turns out that channels A through D are numbers
      // 0 through 3, which we will exploit in the for loop.
      for (int i = 3; i >= 0; i--)
        AD56X4.writeMessage(SS_pin,setMode,i,values[3-i]);
    }
}
void AD56X4Class::setChannel (int SS_pin, byte setMode, word value_D,
                              word value_C, word value_B, word value_A)
{
  word values[] = {value_D,value_C,value_B,value_A};
  AD56X4.setChannel(SS_pin,setMode,values);
}

/* Commands the AD564X DAC whose Slave Select pin is SS_pin to
   update the output (DAC register) of the specified channel from
   its buffer (input register). The valid channel choices are
   
   AD56X4_CHANNEL_A
   AD56X4_CHANNEL_B
   AD56X4_CHANNEL_C
   AD56X4_CHANNEL_D
   AD56X4_CHANNEL_ALL
*/
void AD56X4Class::updateChannel (int SS_pin, byte channel)
{
  AD56X4.writeMessage(SS_pin,AD56X4_COMMAND_UPDATE_DAC_REGISTER,
                            channel,0);
}



/* Commands the AD564X DAC whose Slave Select pin is SS_pin to set
   the given power mode for the specified channels. The power modes
   are
   
   AD56X4_POWERMODE_NORMAL             normal operation (power up)
   AD56X4_POWERMODE_POWERDOWN_1K       connected to ground by 1k
   AD56X4_POWERMODE_POWERDOWN_100K     connected to ground by 100k
   AD56X4_POWERMODE_POWERDOWN_TRISTATE power down in tristate.
   
   A power mode can be applied to a set of channels specified by
   a channel mask (bits 3 through 0 correspond to channels D
   through A), a boolean array (in channel D through A order), or
   four boolean arguments. Or, an array of power modes can be
   applied to each channel (in D through A order).
*/
void AD56X4Class::powerUpDown (int SS_pin, byte powerMode,
                               byte channelMask)
{
  AD56X4.writeMessage(SS_pin,AD56X4_COMMAND_POWER_UPDOWN,0,
                      word((B00110000 & powerMode)
                      | (B00001111 & channelMask)));
}
void AD56X4Class::powerUpDown (int SS_pin, byte powerMode,
                               boolean channels[])
{
  AD56X4.powerUpDown(SS_pin,powerMode,
                     AD56X4.makeChannelMask(channels));
}
void AD56X4Class::powerUpDown (int SS_pin, byte powerMode,
                               boolean channel_D, boolean channel_C,
                               boolean channel_B, boolean channel_A)
{
  AD56X4.powerUpDown(SS_pin,powerMode,
                     AD56X4.makeChannelMask(channel_D,channel_C,
                     channel_B,channel_A));
}
void AD56X4Class::powerUpDown (int SS_pin, byte powerModes[])
{
  // Go through each channel making a mask for just that channel
  // and apply the given power mode.
  
  byte channelMask = 1;
  for (int i = 0; i < 4; i++)
    {
      AD56X4.powerUpDown(SS_pin,powerModes[i],channelMask);
      channelMask = channelMask << 1;
    }
}



/* Commands the AD56X4 DAC whose Slave Select pin is SS_pin to
   reset. The DAC (output) and input (buffer) registers are set
   to zero, and if doing a full reset, the channels are all
   powered up, the external reference is used (internal turned off
   if present), and all channels set so that writing to the input
   register does not auto update the DAC register (output).
*/
void AD56X4Class::reset (int SS_pin, boolean fullReset)
{
  AD56X4.writeMessage(SS_pin,AD56X4_COMMAND_RESET,0, word(fullReset));
}



/* Commands the AD564X DAC whose Slave Select pin is SS_pin which
   channels are to have their DAC register (output) updated
   immediately when the input register (buffer) is set. True is
   for auto update and false is for not. It can either be given
   as a channel mask (bits 3 through 0 correspond to channels D
   through A), a boolean array (in channel D through A order), or
   four boolean arguments.
*/
void AD56X4Class::setInputMode (int SS_pin, byte channelMask)
{
  AD56X4.writeMessage(SS_pin,AD56X4_COMMAND_SET_LDAC,0,
                      word(channelMask));
}
void AD56X4Class::setInputMode (int SS_pin, boolean channels[])
{
  AD56X4.setInputMode(SS_pin,AD56X4.makeChannelMask(channels));
}
void AD56X4Class::setInputMode (int SS_pin, boolean channel_D,
                                boolean channel_C, boolean channel_B,
                                boolean channel_A)
{
  AD56X4.setInputMode(SS_pin,
                      AD56X4.makeChannelMask(channel_D,channel_C,
                                             channel_B,channel_A));
}



/* Commands the AD564X DAC whose Slave Select pin is SS_pin whether
   to use the internal voltage reference or not (use external).
   Should only be used with chips having an internal reference,
   which are the ones whose name ends in an R.
*/
void AD56X4Class::useInternalReference (int SS_pin, boolean yesno)
{
  AD56X4.writeMessage(SS_pin,AD56X4_COMMAND_REFERENCE_ONOFF,0,
                      word(yesno));
}



/* Create channel masks (word with bits 3 through 0 corresponding
   to channels D through A) from an array of booleans for each
   channel (D to A order) or from 4 argument booleans, one for each
   channel.
*/
word AD56X4Class::makeChannelMask (boolean channels[])
{
  return word((byte(channels[0]) << 3) | (byte(channels[1]) << 2)
              | (byte(channels[2]) << 1) | byte(channels[3]));
}
word AD56X4Class::makeChannelMask (boolean channel_D,
                                   boolean channel_C,
                                   boolean channel_B,
                                   boolean channel_A)
{
  return word((byte(channel_D) << 3) | (byte(channel_C) << 2)
              | (byte(channel_B) << 1) | byte(channel_A));
}



/* Writes a 24 bit message to the AD56X4 DAC whose Slave Select
   pin is SS_pin. The message is composed of a command instructing
   the chip what to do, an address telling it which channel/s to
   operate on, and a 2-byte unsigned integer data which could be
   the value to set a channel register to or other control data for
   other commands.
*/
void AD56X4Class::writeMessage (int SS_pin, byte command,
                                byte address, word data)
{
  
  // Set the SPI mode to SPI_MODE1 and the bit order to MSB first.
  
  SPI.setDataMode(SPI_MODE1);
  SPI.setBitOrder(MSBFIRST);
  
  // The clock speed doesn't matter too much as the chip can go up
  // to 50 MHz while the arduino can only go up 8 MHz, so we will
  // leave it.
  
  // Set the Slave Select pin to low so that the DAC knows to
  // listen for a command.
  
  digitalWrite(SS_pin,LOW);
  
  // The first byte is composed of two bits of nothing, then
  // the command bits, and then the address bits. Masks are
  // used for each set of bits and then the fields are OR'ed
  // together.
  
  SPI.transfer((command & B00111000) | (address & B00000111));
  
  // Send the data word. Must be sent byte by byte, MSB first.
  
  SPI.transfer(highByte(data));
  SPI.transfer(lowByte(data));
  
  // Set the Slave Select pin back to high since we are done
  // sending the command.
  
  digitalWrite(SS_pin,HIGH);
  
}


AD56X4.h
/* Copyright (c) 2013, Freja Nordsiek
   All rights reserved.
   
   Redistribution and use in source and binary forms, with or
   without modification, are permitted provided that the
   following conditions are met:
   
   1. Redistributions of source code must retain the above
   copyright notice, this list of conditions and the following disclaimer.
   
   2. Redistributions in binary form must reproduce the above
   copyright notice, this list of conditions and the following
   disclaimer in the documentation and/or other materials
   provided with the distribution.
   
   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
   NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
   LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
   HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
   CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
   OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
   SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

/*
   AD56X4.h: Library for controlling the Analog Devices AD56X4
                 Quad DAC family:
                   AD5624   12-bit
                   AD5664   16-bit
                   AD5624R  12-bit with internal reference
                   AD5644R  14-bit with internal reference
                   AD5664R  16-bit with internal reference
             Data sheet can be found at http://www.analog.com/en/digital-to-analog-converters/da-converters/ad5624r/products/product.html?doc=AD5624R_5644R_5664R.PDF
   
   Author:   Freja Nordsiek
   Notes:
   History:  * 2013-08-15 Created.
*/

/* http://www.analog.com/en/digital-to-analog-converters/da-converters/ad5624r/products/product.html?doc=AD5624R_5644R_5664R.PDF

   The Analog Devices AD56X4 are Quad channel 1X-bit Digital to
   Analog Converter (DAC) controlled by SPI. It can use an external
   reference or its own internal reference if present (an R is
   appended to the chip name) and has various shutdown modes for
   each channel. Each channel has separate DAC (output) and input
   (buffer) registers and commands exist to set just the input
   register, assign the input to the DAC, or set both.
   
   The chip is an SPI slave that expects SPI_MODE1 for its polarity
   and phase (read on falling edge and clock low while idle) that
   can operate at up to 50 MHz, which is far above what the arduino
   can do (8 MHz for the Uno). The DIN, SCLK, and SYNC pins in the
   chip documentation correspond to the Arduino MOSI, SCK, and SS
   pins (doesn't have to use the SS for the arduino, but it is a
   a reasonable default pin). Messages must be sent MSB first and
   are 24 bits. The first two bits are junk and are ignored. The
   next three bits are the command bits, which tell the DAC what
   operation to do. The next three bits are the address bits, which
   tells what channel (or all) should be set. The last two whole
   bytes are a word specifying the value to set the appropriate
   register on the DAC channel if setting a channel, or other needed
   information if doing another sort of command. For the 14-bit and
   12-bit chips, the last 2 and 4 bits respectively are ignored if
   an input or DAC register is being set. The command bits and address
   bits are
   
   Command bits
   | C2 | C1 | C0 | Command
   ------------------------------------------------------------
   | 0  | 0  | 0  | Write to input register N
   | 0  | 0  | 1  | Update DAC register N
   | 0  | 1  | 0  | Write to input register N and update on all
   | 0  | 1  | 1  | Write to and update channel N
   | 1  | 0  | 0  | Power up-down
   | 1  | 0  | 1  | Reset
   | 1  | 1  | 0  | Set LDAC register
   | 1  | 1  | 1  | Internal reference on/off (if present)
   ------------------------------------------------------------
   
   Address bits
   | A2 | A1 | A0 | Channel/s
   --------------------------
   | 0  | 0  | 0  | A
   | 0  | 0  | 1  | B
   | 0  | 1  | 0  | C
   | 0  | 1  | 1  | D
   | 1  | 1  | 1  | All
   --------------------------
   
   For the power up-down command, the power mode specified with
   data bits DB5 and DB4 are applied to the channels specified by
   ones in the bit mask made by DB3 to DB0 (in channel D to A
   order). The power modes are
   
   Power up-down mode
   | DB5 | DB4 | mode
   ------------------------------
   |  0  |  0  | normal operation
   |  0  |  1  | 1k to ground
   |  1  |  0  | 100k to ground
   |  1  |  1  | tri-state
   ------------------------------
   
   For the reset command, DB0 controls exactly what is reset.
   Regardless of its value, the input and DAC registers are all
   reset to zero. If DB0 is set to one; then the LDAC register
   is reset to zero, all channels powered up to normal mode, and
   the internal reference turned off (means external reference is
   used).
   
   For the set LDAC register command, whether to just set the DAC
   register to the input register every time it is changed or not,
   is set for each channel with DB3 to DB0 corresponding to channel
   D through A. A zero means that setting the input register does
   not automatically set the DAC register (value isn't written out
   as output yet) and a one means that the DAC register (and hence
   the output) is set the the input register the moment the input
   register is changed.
   
   For the internal reference on/off command, setting DB0 to zero
   turns it off and setting it to one turns it on. When off, the
   external reference is used as the voltage reference. This
   command only works in the R versions that have an internal
   reference.
   
*/

#ifndef AD56X4_h
#define AD56X4_h

#include "Arduino.h"
#include <SPI.h>

/* For the various defined values, they are chosen so that they
   are exactly the values that need to be put into the SPI
   message. This makes it a lot easier to construct the message.
*/

#define AD56X4_COMMAND_WRITE_INPUT_REGISTER            B00000000
#define AD56X4_COMMAND_UPDATE_DAC_REGISTER             B00001000
#define AD56X4_COMMAND_WRITE_INPUT_REGISTER_UPDATE_ALL B00010000
#define AD56X4_COMMAND_WRITE_UPDATE_CHANNEL            B00011000
#define AD56X4_COMMAND_POWER_UPDOWN                    B00100000
#define AD56X4_COMMAND_RESET                           B00101000
#define AD56X4_COMMAND_SET_LDAC                        B00110000
#define AD56X4_COMMAND_REFERENCE_ONOFF                 B00111000

#define AD56X4_CHANNEL_A                               B00000000
#define AD56X4_CHANNEL_B                               B00000001
#define AD56X4_CHANNEL_C                               B00000010
#define AD56X4_CHANNEL_D                               B00000011
#define AD56X4_CHANNEL_ALL                             B00000111

#define AD56X4_SETMODE_INPUT                           AD56X4_COMMAND_WRITE_INPUT_REGISTER
#define AD56X4_SETMODE_INPUT_DAC                       AD56X4_COMMAND_WRITE_UPDATE_CHANNEL
#define AD56X4_SETMODE_INPUT_DAC_ALL                   AD56X4_COMMAND_WRITE_INPUT_REGISTER_UPDATE_ALL

#define AD56X4_POWERMODE_NORMAL                        B00000000
#define AD56X4_POWERMODE_POWERDOWN_1K                  B00010000
#define AD56X4_POWERMODE_POWERDOWN_100K                B00100000
#define AD56X4_POWERMODE_TRISTATE                      B00110000



class AD56X4Class
{
  
  public:
  
    static void setChannel (int SS_pin, byte setMode, byte channel,
                            word value);
    static void setChannel (int SS_pin, byte setMode, word values[]);
    static void setChannel (int SS_pin, byte setMode, word value_D,
                            word value_C, word value_B, word value_A);
                            
    static void updateChannel (int SS_pin, byte channel);
    
    static void powerUpDown (int SS_pin, byte powerMode,
                             boolean channels[]);
    static void powerUpDown (int SS_pin, byte powerMode,
                             boolean channel_D, boolean channel_C,
                             boolean channel_B, boolean channel_A);
    static void powerUpDown (int SS_pin, byte powerModes[]);
    
    static void reset (int SS_pin, boolean fullReset);
    
    static void setInputMode (int SS_pin, boolean channels[]);
    static void setInputMode (int SS_pin, boolean channel_D,
                              boolean channel_C, boolean channel_B,
                              boolean channel_A);
    
    static void useInternalReference (int SS_pin, boolean yesno);
    
  private:
    inline static void powerUpDown (int SS_pin, byte powerMode,
                                    byte channelMask);
    
    inline static void setInputMode (int SS_pin, byte channelMask);
    
    static word makeChannelMask (boolean channels[]);
    static word makeChannelMask (boolean channel_D, boolean channel_C,
                                 boolean channel_B, boolean channel_A);
    
    static void writeMessage (int SS_pin, byte command, byte address,
                              word data);
    
};

extern AD56X4Class AD56X4;

#endif 
